photon-core/src/main/java/org/photonvision/vision/pipe/impl/MultiTargetPNPPipe.java

```diff
@@ -0,0 +1,106 @@
+import edu.wpi.first.math.geometry.Pose3d;
+import java.util.ArrayList;
+import java.util.List;
+import org.opencv.core.Mat;
+import org.photonvision.common.logging.LogGroup;
+import org.photonvision.common.logging.Logger;
+import org.photonvision.estimation.VisionEstimation;
+import org.photonvision.targeting.MultiTargetPNPResults;
+import org.photonvision.targeting.TargetCorner;
+import org.photonvision.vision.calibration.CameraCalibrationCoefficients;
+import org.photonvision.vision.pipe.CVPipe;
+import org.photonvision.vision.target.TrackedTarget;
+
+/** Estimate the camera pose given multiple Apriltag observations */
+public class MultiTargetPNPPipe
+        extends CVPipe<
+                List<TrackedTarget>, MultiTargetPNPResults, MultiTargetPNPPipe.MultiTargetPNPPipeParams> {
+    private static final Logger logger = new Logger(MultiTargetPNPPipe.class, LogGroup.VisionModule);
+
+    private boolean hasWarned = false;
+
+    @Override
+    protected MultiTargetPNPResults process(List<TrackedTarget> targetList) {
+        if (params.cameraCoefficients == null
+                || params.cameraCoefficients.getCameraIntrinsicsMat() == null
+                || params.cameraCoefficients.getDistCoeffsMat() == null) {
+            if (!hasWarned) {
+                logger.warn(
+                        "Cannot perform solvePNP an uncalibrated camera! Please calibrate this resolution...");
+                hasWarned = true;
+            }
+            return new MultiTargetPNPResults();
+        }
+
+        return calculateTargetPose(targetList);
+    }
+
+    private MultiTargetPNPResults calculateTargetPose(List<TrackedTarget> targetList) {
+        if (params.cameraCoefficients == null
+                || params.cameraCoefficients.getCameraIntrinsicsMat() == null
+                || params.cameraCoefficients.getDistCoeffsMat() == null) {
+            return new MultiTargetPNPResults();
+        }
+
+        var visCorners = new ArrayList<TargetCorner>();
+        var knownVisTags = new ArrayList<AprilTag>();
+        var tagIDsUsed = new ArrayList<Integer>();
+        for (var target : targetList) {
+            for (var corner : target.getTargetCorners()) {
+                visCorners.add(new TargetCorner(corner.x, corner.y));
+            }
+            Pose3d tagPose = params.atfl.getTagPose(target.getFiducialId()).get();
+
+            // actual layout poses of visible tags -- not exposed, so have to recreate
+            knownVisTags.add(new AprilTag(target.getFiducialId(), tagPose));
+            tagIDsUsed.add(target.getFiducialId());
+        }
+        var ret = new MultiTargetPNPResults();
+        ret.estimatedPose =
+                VisionEstimation.estimateCamPosePNP(
+                        params.cameraCoefficients.getCameraIntrinsicsMat(),
+                        params.cameraCoefficients.getDistCoeffsMat(),
+                        visCorners,
+                        knownVisTags);
```

Comment:
Why not run with just one target? I guess we just want user code to catch that? Like there isn't a technical reason why one target just works, it's just less good?

Reply:
If there is only one target, it should be handled by `AprilTagPoseEstimatorPipe` instead. If only one target is passed to this `estimateCamPosePNP()` method, it would use the SOLVEPNP_IPPE_SQUARE solver to redundantly estimate the camera-to-tag transform and camera pose, and then provide a non-empty multitag result using that. We want `PhotonPoseEstimator` to see an empty multitag result (since there is only one tag) and go to the fallback strategy, which users can see reflected in their `EstimatedRobotPose`.

Reply:
gotcha yeah. done